### DISCUSSION TITLE
setting providers

### DISCUSSION THREAD

**Discussion by leksmano** [logfellow/logstash-logback-encoder#990] [Q&A] [answered]
I noticed that this PR has removed the ability to set providers (going from 7.2 to 7.4):
https://github.com/logfellow/logstash-logback-encoder/pull/861

However documentation still appears to have examples referencing setting providers:
https://github.com/logfellow/logstash-logback-encoder/blob/b1e7653914fba93baf6042b15d44e54909879e4f/stack-hash.md?plain=1#L292

I am guessing this configuration is no longer applicable?

Also for setting the ShortenedThrowableConverter we likely don't need to set them under a provider
https://github.com/logfellow/logstash-logback-encoder/blob/b1e7653914fba93baf6042b15d44e54909879e4f/src/test/resources/logback-test.xml#L42

**Comment by philsttr** [comment, marked as answer]
### LoggingEventCompositeJsonEncoder

`LoggingEventCompositeJsonEncoder` allows the `<providers>` element.  And `LoggingEventCompositeJsonEncoder` is what is being used in the documentation you referenced.

If you want to use the `ShortenedThrowableConverter` with a `LoggingEventCompositeJsonEncoder`, then you add the `stackTrace` provider with a `throwableConverter` element.

```java
<encoder class="net.logstash.logback.encoder.LoggingEventCompositeJsonEncoder">
    <providers>
        ... snip ... other providers
        <stackTrace>
            <throwableConverter class="net.logstash.logback.stacktrace.ShortenedThrowableConverter">
                ... snip ... ShortenedThrowableConverter configuration
            </throwableConverter>
        </stackTrace>
    </providers>
</encoder>
```

### LogstashEncoder

`LogstashEncoder` does not allow the `<providers>` element.  It is a configuration error to do so.  What changed in #861 is how that error is reported.  Prior to 7.3, an error would be output via logback's status manager, which could be seen in [status listeners](https://logback.qos.ch/manual/configuration.html#dumpingStatusData).  Starting with 7.3 an exception is thrown instead.

If you want to add additional providers to `LogstashEncoder`, use the `<provider class="....">` element as described in [Registering Additional Providers](https://github.com/logfellow/logstash-logback-encoder#registering-additional-providers).

If you want to use the `ShortenedThrowableConverter` with a `LogstashEncoder`, then add the `throwableConverter` element in the encoder.

```
<encoder class="net.logstash.logback.encoder.LogstashEncoder">
    <throwableConverter class="net.logstash.logback.stacktrace.ShortenedThrowableConverter">
        ... snip ... ShortenedThrowableConverter configuration
    </throwableConverter>
</encoder>
```
